File: julia/converter.jl
# To run: julia converter.jl ../input.md
# Pros:
#   - Multi-methods are neat!
#   - File IO is super easy
#   - Really clean syntax
#   - String interpolation
#   - Expression function syntax
# Cons:
#   - Arrays are 1-indexed

abstract type Node end

struct Header <: Node
    size::Int
    children::Vector{Node}
end

getHTML(header::Header) = "<h$(header.size)>$(join(map(child->getHTML(child), header.children), ""))</h$(header.size)>\n"

struct Paragraph <: Node
    children::Vector{Node}
end

getHTML(paragraph::Paragraph) = "<p>$(join(map(child->getHTML(child), paragraph.children), ""))</p>\n\n"

struct CodeBlock <: Node
    text::String
end

getHTML(codeblock::CodeBlock) = "<pre><code>$(codeblock.text)</code></pre>\n\n"

struct Image <: Node
    text::String
    url::String
end

getHTML(image::Image) = "<img src=\"$(image.url)\" alt=\"$(image.text)\" />"

struct Text <: Node
    text::String
end

getHTML(text::Text) = "$(text.text)"

struct Italic <: Node
    children::Vector{Node}
end

getHTML(italic::Italic) = "<em>$(join(map(child->getHTML(child), italic.children), ""))</em>"

struct Bold <: Node
    children::Vector{Node}
end

getHTML(bold::Bold) = "<strong>$(join(map(child->getHTML(child), bold.children), ""))</strong>"

struct Code <: Node
    text::String
end

getHTML(code::Code) = "<code>$(code.text)</code>"

struct Link <: Node
    text::String
    url::String
end

getHTML(link::Link) = "<a href=\"$(link.url)\">$(link.text)</a>"

struct Token
    data::String
    line::Int
    col::Int
end

mutable struct Parser
    index::Int
    tokens::Vector{Token}
end

function createParser(contents::String)
    parser::Parser = Parser(1, [])

    data::String = SubString(contents, 1, 1)
    line::Int = 1
    col::Int = 1
    oldCol::Int = 1
    oldC::Char = data[1]
    for c::Char in SubString(contents, 2)
        if oldC == '\n' || oldC == '\r' || (isSpecialChar(oldC) != isSpecialChar(c)) || contains("#[(!\n\r", string(c))
            if !contains(data, "\r")
                push!(parser.tokens, Token(data, line, oldCol))
            end
            oldCol = col + 1
            if c == '\n'
                line += 1
                col = 0
            end
            data = ""
        end
        if oldC != '#' || !isspace(c)
            data = string(data, c)
        end
        col += 1
        oldC = c
    end
    push!(parser.tokens, Token(data, line, col))
    push!(parser.tokens, Token("\n", line, col))

    parser
end

isSpecialChar(c::Char) = contains("_*`#[]()!", string(c))

function pop(parser::Parser)
    parser.index += 1
    parser.tokens[parser.index - 1]
end

peek(parser::Parser) = parser.tokens[parser.index]

accept(parser::Parser, data::String) =
    if peek(parser).data == data
        pop(parser)
    else
        nothing
    end

expect(parser::Parser, data::String) =
    if accept(parser, data) === nothing
        top = peek(parser)
        if isSpecialChar(top.data[1])
            println("error: $(top.line):$(top.col) expected `$(data)` got `$(top.data)`")
        else
            println("error: $(top.line):$(top.col) expected `$(data)` got text")
        end
        exit(1)
    end


takeUntil(parser::Parser, sentinel::String) = 
    if accept(parser, sentinel) === nothing
        pop(parser).data * takeUntil(parser, sentinel)
    else
        ""
    end

parseDocument(parser::Parser) =
    if parser.index < length(parser.tokens) - 1
        node = parseNode(parser)
        rest = parseDocument(parser)
        if node !== nothing 
            push!(rest, node)
        end
        rest
    else
        []
    end

parseNode(parser::Parser) =
    if accept(parser, "#") !== nothing
        parseHeader(parser)
    elseif accept(parser, "```") !== nothing
        parseCodeBlock(parser)
    elseif accept(parser, "!") !== nothing
        parseImage(parser)
    elseif accept(parser, "\n") === nothing
        parseParagraph(parser)
    else
        nothing
    end

function parseHeader(parser::Parser)
    size = 1
    while accept(parser, "#") !== nothing
        size += 1
    end
    Header(size, parseFormattedText(parser, ["\n"]))
end

parseParagraph(parser::Parser) = Paragraph(parseFormattedText(parser, ["\n"]))

parseCodeBlock(parser::Parser) = CodeBlock(takeUntil(parser, "```"))

function parseImage(parser::Parser)
    expect(parser, "[")
    text = pop(parser).data
    expect(parser, "]")
    expect(parser, "(")
    url = pop(parser).data
    expect(parser, ")")
    Image(text, url)
end

function parseFormattedText(parser::Parser, bounds::Vector{String})
    retval = []
    while !issubset([peek(parser).data], bounds)
        if accept(parser, "_") !== nothing || accept(parser, "*") !== nothing
            push!(retval, parseItalic(parser, copy(bounds)))
        elseif accept(parser, "__") !== nothing || accept(parser, "**") !== nothing
            push!(retval, parseBold(parser, copy(bounds)))
        elseif accept(parser, "`") !== nothing
            push!(retval, parseCode(parser))
        elseif accept(parser, "[") !== nothing
            push!(retval, parseLink(parser))
        else
            push!(retval, Text(pop(parser).data))
        end 
    end
    retval
end

function parseItalic(parser::Parser, bounds::Vector{String})
    push!(bounds, "*")
    push!(bounds, "_")
    children = parseFormattedText(parser, bounds)
    if accept(parser, "*") === nothing
        expect(parser, "_")
    end
    Italic(children)
end

function parseBold(parser::Parser, bounds::Vector{String})
    push!(bounds, "**")
    push!(bounds, "__")
    children = parseFormattedText(parser, bounds)
    if accept(parser, "**") === nothing
        expect(parser, "__")
    end
    Bold(children)
end

parseCode(parser::Parser) = Code(takeUntil(parser, "`"))

function parseLink(parser::Parser)
    text = pop(parser).data
    expect(parser, "]")
    expect(parser, "(")
    url = pop(parser).data
    expect(parser, ")")
    Link(text, url)
end

if length(ARGS) != 1
    println("usage: julia converter.jl <markdown-filename>")
else
    # Open input file, read contents
    f = open(ARGS[1])
    output = open("output.html", "w")
    write(output, "$(join(map(node->getHTML(node), reverse(parseDocument(createParser(read(f, String))))), ""))")
    close(output)
    close(f)
end

File: nim/converter.nim
# To run: nim compile --run converter.nim -- ../input.md
# Pros:
#   - Pretty terse and simple
#   - Strings are fairly easy
#   - I like the UFCS style of OOP, its honestly not too bad!
#   - Immutability
#   - Function body is result
#   - Generics work fairly well
#   - Able to chain `type` and probably `let`, `var`, `const`s, pretty neat!
#   - Structs initializers are good
# Cons:
#   - Weird data model, with sequences and stuff, but no access to lower level stuff
#   - Have to define procs in order. Maybe just a lang-server thing?
#   - Identifiers are just undefined if theyre not defined. Id rather have an error, when is that behavior ever desired?
#   - Overloading causes strange error messages, like in C++
#   - Have to import options, instead of them just being there
import os
import std/options
import std/strformat

type
    Node = ref object of RootObj
    Header = ref object of Node
        size: int
        children: seq[Node]
    Paragraph = ref object of Node
        children: seq[Node]
    CodeBlock = ref object of Node
        text: string
    Image = ref object of Node
        text: string
        url: string
    Text = ref object of Node
        text: string
    Italic = ref object of Node
        children: seq[Node]
    Bold = ref object of Node
        children: seq[Node]
    Code = ref object of Node
        text: string
    Link = ref object of Node
        text: string
        url: string
    Token = object
        data: string
        line: int
        col: int
    Parser = object
        tokens: seq[Token]
        index: int

method get_HTML(node: Node): string {.base.} =
    quit "to override!"

method get_HTML(header: Header): string =
    var children_string = ""
    for child in header.children:
        children_string.add(child.get_HTML())
    fmt"<h{header.size}>{children_string}</h{header.size}>" & "\n\n"

method get_HTML(paragraph: Paragraph): string =
    var children_string = ""
    for child in paragraph.children:
        children_string.add(child.get_HTML())
    fmt"<p>{children_string}</p>" & "\n\n"

method get_HTML(code_block: CodeBlock): string = 
    fmt"<pre><code>{code_block.text}</code></pre>" & "\n\n"

method get_HTML(image: Image): string = 
    fmt"<img src='{image.url}' alt='{image.text}' />"

method get_HTML(text: Text): string = text.text

method get_HTML(italic: Italic): string =
    var children_string = ""
    for child in italic.children:
        children_string.add(child.get_HTML())
    fmt"<em>{children_string}</em>"

method get_HTML(bold: Bold): string =
    var children_string = ""
    for child in bold.children:
        children_string.add(child.get_HTML())
    fmt"<strong>{children_string}</strong>"

method get_HTML(code: Code): string = fmt"<code>{code.text}</code>"

method get_HTML(link: Link): string = 
    fmt"<a href='{link.url}'>{link.text}</a>"

proc is_special_char(c: char): bool = 
    c in @['_', '*', '`', '#', '[', ']', '(', ')', '!']

proc tokenize(parser: var Parser, contents: string) =
    var data = ""
    data.add(contents[0])
    var line = 1
    var col = 1
    var old_col = 1
    var old_c = contents[0]
    for c in contents[1 ..< len(contents)]:
        if old_c == '\n' or old_c == '\r' or (is_special_char(old_c) != is_special_char(c)) or c in @['#', '[', '(', '!', '\n', '\r']:
            parser.tokens.add(Token(data: data, line: line, col: old_col))
            old_col = col + 1
            if c == '\n':
                line += 1
                col = 0
            data = ""
        if old_c != '#' or not (c in @[' ', '\t', '\n', '\r']):
            data.add(c)
        col += 1
        old_c = c
    parser.tokens.add(Token(data: data, line: line, col: col))
    parser.tokens.add(Token(data: "\n", line: line, col: col))

proc pop(parser: var Parser): Token =
    parser.index += 1
    parser.tokens[parser.index - 1]

proc peek(parser: var Parser): Token = parser.tokens[parser.index]

proc accept(parser: var Parser, data: string): Option[string] = 
    if parser.peek().data == data:
        some(parser.pop().data)
    else:
        none(string)

proc expect(parser: var Parser, data: string) =
    if parser.accept(data).isNone:
        var top = parser.peek()
        if is_special_char(top.data[0]):
            echo fmt"error: {top.line}:{top.col} expected `{data}` got `{top.data}`"
        else:
            echo fmt"error: {top.line}:{top.col} expected `{data}` text"
        system.quit 1
        
proc take_until(parser: var Parser, sentinel: string): string =
    var text = ""
    while parser.accept(sentinel).isNone:
        text = text & parser.pop().data
    text

proc parse_formatted_text(parser: var Parser, bounds: seq[string]): seq[Node]

proc parse_italic(parser: var Parser, bounds: seq[string]): Node =
    var new_bounds = newSeq[string](0)
    for bound in bounds:
        new_bounds.add(bound)
    new_bounds.add("*")
    new_bounds.add("_")
    var children = parser.parse_formatted_text(new_bounds)
    if parser.accept("*").isNone:
        parser.expect("_")
    return Italic(children: children)

proc parse_bold(parser: var Parser, bounds: seq[string]): Node =
    var new_bounds = newSeq[string](0)
    for bound in bounds:
        new_bounds.add(bound)
    new_bounds.add("**")
    new_bounds.add("__")
    var children = parser.parse_formatted_text(new_bounds)
    if parser.accept("**").isNone:
        parser.expect("__")
    return Bold(children: children)

proc parse_code(parser: var Parser): Node = Code(text: parser.take_until("`"))

proc parse_link(parser: var Parser): Node = 
    var text = parser.pop().data
    parser.expect("]")
    parser.expect("(")
    var url = parser.pop().data
    parser.expect(")")
    Link(text: text, url: url)

proc parse_formatted_text(parser: var Parser, bounds: seq[string]): seq[Node] =
    var retval = newSeq[Node](0)
    while parser.peek().data notin bounds:
        if parser.accept("_").isSome or parser.accept("*").isSome:
            retval.add(parser.parse_italic(bounds))
        elif parser.accept("__").isSome or parser.accept("**").isSome:
            retval.add(parser.parse_bold(bounds))
        elif parser.accept("`").isSome:
            retval.add(parser.parse_code())
        elif parser.accept("[").isSome:
            retval.add(parser.parse_link())
        else:
            retval.add(Text(text: parser.pop().data))
    return retval

proc parse_header(parser: var Parser): Node = 
    var size = 1
    while parser.accept("#").isSome:
        size += 1
    Header(size: size, children: parser.parse_formatted_text(@["\n"]))

proc parse_code_block(parser: var Parser): Node = 
    CodeBlock(text: parser.take_until("```"))

proc parse_image(parser: var Parser): Node = 
    parser.expect("[")
    var text = parser.pop().data
    parser.expect("]")
    parser.expect("(")
    var url = parser.pop().data
    parser.expect(")")
    Image(text: text, url: url)

proc parse_paragraph(parser: var Parser): Node = 
    Paragraph(children: parser.parse_formatted_text(@["\n"]))

proc parse_node(parser: var Parser): Option[Node] = 
    if parser.accept("#").isSome:
        some(parser.parse_header())
    elif parser.accept("```").isSome:
        some(parser.parse_code_block())
    elif parser.accept("!").isSome:
        some(parser.parse_image())
    elif parser.accept("\n").isSome:
        none(Node)
    else:
        some(parser.parse_paragraph())

proc parse_document(parser: var Parser): seq[Node] =
    var retval = newSeq[Node](0)
    while parser.index < len(parser.tokens) - 1:
        var node = parser.parse_node()
        if node.isSome:
            retval.add(node.get())
    return retval

if paramCount() != 2:
    quit "usage: ./converter <markdown-filename>"
else:
    # Open contents, parse
    var parser: Parser = Parser(tokens: newSeq[Token](0), index: 0)
    parser.tokenize(readFile(paramStr(2)))
    var document = parser.parse_document()
    # Write output
    var outstring = ""
    for node in document:
        outstring.add(node.get_HTML())
    writeFile("output.html", outstring)